### PR TITLE
KAFKA-17561: KIP-1091 add operator metrics

### PR DESCRIPTION
**Pull request by bbejeck**
Implementation of KIP-1091 adding operator metrics to Kafka Streams

Updated existing tests to validate added metrics

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java [hunk 1]
```diff
@@ -180,15 +180,16 @@ public void shouldPushMetricsToBroker(final String recordingLevel) throws Except
                         final String name = mn.name().replace('-', '.');
                         final String group = mn.group().replace("-metrics", "").replace('-', '.');
                         return "org.apache.kafka." + group + "." + name;
-                    }).sorted().collect(Collectors.toList());
+                    }).filter(name -> !name.equals("org.apache.kafka.stream.thread.state"))// telemetry reporter filters out string metrics
+                    .sorted().collect(Collectors.toList());
```
Comment (bbejeck) [inline streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:184]: Telemetry filters out string metrics, so I remove it here from the expected metrics list

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -1250,6 +1253,20 @@ private Optional<String> removeStreamThread(final long timeoutMs) throws Timeout
...
 
+    private int calculateMetricsRecordingLevel() {
+        final int recordingLevel;
+        final String recordingLevelString = applicationConfigs.getString(METRICS_RECORDING_LEVEL_CONFIG);
+        if (recordingLevelString.equals("INFO")) {
+            recordingLevel = 0;
+        } else if (recordingLevelString.equals("DEBUG")) {
+            recordingLevel = 1;
+        } else {
+            // Must be TRACE level
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Might be better to use another `if` and final `else` throw an exception as safe guard? Would also highlight that we need to update this code, in case we add a new level?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 3]
```diff
@@ -1250,6 +1253,20 @@ private Optional<String> removeStreamThread(final long timeoutMs) throws Timeout
         return Optional.empty();
     }
 
+    private int calculateMetricsRecordingLevel() {
+        final int recordingLevel;
+        final String recordingLevelString = applicationConfigs.getString(METRICS_RECORDING_LEVEL_CONFIG);
+        if (recordingLevelString.equals("INFO")) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why not use a `switch` statement?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -614,6 +615,12 @@ public StreamThread(final Time time,
             streamsMetrics,
             time.milliseconds()
         );
+        ThreadMetrics.addThreadStateTelemetryMetric(threadId,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: `threadId` should be in the next line by itself

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -614,6 +615,12 @@ public StreamThread(final Time time,
...
+                streamsMetrics,
+                (metricConfig, now) -> this.state().ordinal());
+        ThreadMetrics.addThreadStateMetric(threadId,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -614,6 +615,12 @@ public StreamThread(final Time time,
...
+                streamsMetrics,
+                (metricConfig, now) -> this.state().name().toLowerCase(Locale.getDefault()));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: For the corresponding client metric we just use
```
 ClientMetrics.addStateMetric(streamsMetrics, (metricsConfig, now) -> state);
 ```
 
 Why so "complicated" -- I am also ok to update the code for the client metric. But both should be the same?
 
 (Or maybe keep `... -> state` and add a "fancy" `toString()` overload to both `enum` (for client and thread) which model the state?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java [hunk 7]
```diff
@@ -290,6 +295,30 @@ public static void addThreadStartTimeMetric(final String threadId,
         );
     }
 
+    public static void addThreadStateTelemetryMetric(final String threadId,
+                                                     final StreamsMetricsImpl streamsMetrics,
+                                                     final Gauge<Integer> threadStateProvider) {
+        streamsMetrics.addThreadLevelMutableMetric(
+                THREAD_STATE,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java]: nit: indention too deep (should only be 4 whitespace, not 8 -- should be an IDE setting)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java [hunk 8]
```diff
@@ -290,6 +295,30 @@ public static void addThreadStartTimeMetric(final String threadId,
...
+                threadId,
+                threadStateProvider
+        );
+    }
+
+    public static void addThreadStateMetric(final String threadId,
+                                            final StreamsMetricsImpl streamsMetrics,
+                                            final Gauge<String> threadStateProvider) {
+        streamsMetrics.addThreadLevelMutableMetric(
+                STATE,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java]: as above

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java [hunk 9]
```diff
@@ -180,15 +180,16 @@ public void shouldPushMetricsToBroker(final String recordingLevel) throws Except
...
             final List<String> actualMetrics = new ArrayList<>(TelemetryPlugin.SUBSCRIBED_METRICS.get(mainConsumerInstanceId));
             assertEquals(expectedMetrics, actualMetrics);
 
             TestUtils.waitForCondition(() -> !TelemetryPlugin.SUBSCRIBED_METRICS.get(adminInstanceId).isEmpty(),
                     30_000,
                     "Never received subscribed metrics");
             final List<String> actualInstanceMetrics = TelemetryPlugin.SUBSCRIBED_METRICS.get(adminInstanceId);
-            final List<String> expectedInstanceMetrics = Arrays.asList("org.apache.kafka.stream.alive.stream.threads", "org.apache.kafka.stream.failed.stream.threads");
+            final List<String> expectedInstanceMetrics = Arrays.asList("org.apache.kafka.stream.alive.stream.threads", "org.apache.kafka.stream.client.state", "org.apache.kafka.stream.failed.stream.threads", "org.apache.kafka.stream.recording.level");
```
Comment (mjsax) [inline streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java]: Nit: line too long. How about
```
final List<String> expectedInstanceMetrics = Arrays.asList(
    "org.apache.kafka.stream.client.state",
    "org.apache.kafka.stream.alive.stream.threads",
    "org.apache.kafka.stream.failed.stream.threads",
    "org.apache.kafka.stream.recording.level"
);
```

**Review by mjsax — COMMENTED** [review]
Made a pass. Might be good if @cadonna could take a look too.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java [hunk 10]
```diff
@@ -413,6 +415,39 @@ public void shouldAddThreadStartTimeMetric() {
         );
     }
 
+    @Test
+    public void shouldAddThreadStateTelemetryMetric() {
+        final Gauge<Integer> threadStateProvider = (streamsMetrics, startTime) -> StreamThread.State.RUNNING.ordinal();
+        ThreadMetrics.addThreadStateTelemetryMetric(
+                "threadId",
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java]: There is a constant for a test thread ID:

```suggestion
                THREAD_ID,
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java [hunk 11]
```diff
@@ -413,6 +415,39 @@ public void shouldAddThreadStartTimeMetric() {
...
+                streamsMetrics,
+                threadStateProvider
+        );
+        verify(streamsMetrics).addThreadLevelMutableMetric(
+                "thread-state",
+                "The current state of the thread",
+                "threadId",
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java]:
```suggestion
                THREAD_ID,
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java [hunk 12]
```diff
@@ -413,6 +415,39 @@ public void shouldAddThreadStartTimeMetric() {
...
+                threadStateProvider
+        );
+    }
+
+    @Test
+    public void shouldAddThreadStateJMXMetric() {
+        final Gauge<String> threadStateProvider = (streamsMetrics, startTime) -> StreamThread.State.RUNNING.name().toLowerCase(Locale.getDefault());
+        ThreadMetrics.addThreadStateMetric(
+                "threadId",
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java]:
```suggestion
                THREAD_ID,
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java [hunk 13]
```diff
@@ -413,6 +415,39 @@ public void shouldAddThreadStartTimeMetric() {
...
+                streamsMetrics,
+                threadStateProvider
+        );
+        verify(streamsMetrics).addThreadLevelMutableMetric(
+                "state",
+                "The current state of the thread",
+                "threadId",
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java]:
```suggestion
                THREAD_ID,
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 4 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: we usually use 4 spaces indentation.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 14]
```diff
@@ -1067,7 +1070,7 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
     private static Metrics createMetrics(final StreamsConfig config, final Time time, final String clientId) {
         final MetricConfig metricConfig = new MetricConfig()
             .samples(config.getInt(StreamsConfig.METRICS_NUM_SAMPLES_CONFIG))
-            .recordLevel(Sensor.RecordingLevel.forName(config.getString(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG)))
+            .recordLevel(Sensor.RecordingLevel.forName(config.getString(METRICS_RECORDING_LEVEL_CONFIG)))
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why did you remove the `StreamsConfig.` prefix? For all other config names, we use the prefix.

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java (hunk 1 above)
Comment (cadonna) [reply streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:184]: I do not understand this. According to the KIP and the code in `ThreadMetrics` the type of the metric is numeric.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java [hunk 15]
```diff
@@ -413,6 +415,39 @@ public void shouldAddThreadStartTimeMetric() {
...
+        verify(streamsMetrics).addThreadLevelMutableMetric(
+                "thread-state",
+                "The current state of the thread",
+                "threadId",
+                threadStateProvider
+        );
+    }
+
+    @Test
+    public void shouldAddThreadStateJMXMetric() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetricsTest.java]: nit:
```suggestion
    public void shouldAddThreadStateJmxMetric() {
```

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @bbejeck !

Could you please also update `MetricsIntegrationTest` with the new metrics?

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java (hunk 1 above)
Comment (bbejeck) [reply streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:184]: We also added a  `state` metric for stream threads that reports the current state of the thread as a string.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 3 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I thought [Java 11 had switch expressions](https://docs.oracle.com/en/java/javase/13/language/switch-expressions.html) but it's Java 13+ so I was a little disappointed, but now looking at again the code will still benefit from using `switch`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 6 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: The KIP specifies lower-case string but I like the other approach so I'll update it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java (hunk 7 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java]: Ack, I think I need to copy settings from someone this is a common occurance with my PRs

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 14 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Unintended, I let IntelliJ do the driving on that one, I'll revert

**Comment by bbejeck** [comment]
@mjsax and @cadonna - thanks for the reviews! All comments addressed.

**Review by mjsax — APPROVED** [review]
LGTM. Assuming build passes.

File: streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java (hunk 1 above)
Comment (cadonna) [reply streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:184]: Got it! 
I thought `org.apache.kafka.stream.thread.thread.state` was a typo and it should actually be `org.apache.kafka.stream.thread.state`. So I was wondering, why you wanted to exclude the metric you explicitly added for sending it to the broker from the test.

**Review by cadonna — APPROVED** [review]
Thanks for the updates, @bbejeck !

LGTM!

**Comment by bbejeck** [comment]
Merged #17820 into trunk
